cache/demux.go

```diff
@@ -0,0 +1,130 @@
+package cache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type demux struct {
+	mu sync.RWMutex
+	// activeWatchers & laggingWatchers hold the first revision the watcher still needs (nextRev).
+	activeWatchers  map[*watcher]int64
+	laggingWatchers map[*watcher]int64
+	history         *ringBuffer
+	resyncInterval  time.Duration
+}
+
+func newDemux(ctx context.Context, wg *sync.WaitGroup, history *ringBuffer, resyncInterval time.Duration) *demux {
+	d := &demux{
+		activeWatchers:  make(map[*watcher]int64),
+		laggingWatchers: make(map[*watcher]int64),
+		history:         history,
+		resyncInterval:  resyncInterval,
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		d.resyncLoop(ctx)
+	}()
+	return d
+}
+
+func (d *demux) add(w *watcher, nextRev int64) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	var latest int64
+	if latestEvent := d.history.PeekLatest(); latestEvent != nil {
+		latest = latestEvent.Kv.ModRevision
+	}
+	if nextRev <= latest {
+		d.laggingWatchers[w] = nextRev
+	} else {
+		d.activeWatchers[w] = nextRev
+	}
+}
+
+func (d *demux) remove(w *watcher) {
+	func() {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		delete(d.activeWatchers, w)
+		delete(d.laggingWatchers, w)
+	}()
+	w.Stop()
+}
+
+func (d *demux) broadcast(event *clientv3.Event) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	for w, nextRev := range d.activeWatchers {
+		if !w.enqueueEvent(event, &nextRev) { // buffer overflow → fell behind
+			nextRev = event.Kv.ModRevision
+			delete(d.activeWatchers, w)
+			d.laggingWatchers[w] = nextRev
+		} else {
+			d.activeWatchers[w] = nextRev
+		}
+	}
+}
+
+// Called when etcd compaction invalidates our cached history, so clients should resubscribe.
+func (d *demux) purgeAll() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	for w := range d.activeWatchers {
+		w.Stop()
+	}
+	for w := range d.laggingWatchers {
+		w.Stop()
+	}
+	d.activeWatchers = make(map[*watcher]int64)
+	d.laggingWatchers = make(map[*watcher]int64)
+}
+
+// resyncLoop periodically tries to catch lagging watchers up by replaying events from History.
+func (d *demux) resyncLoop(ctx context.Context) {
+	timer := time.NewTimer(d.resyncInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			d.resyncLaggingWatchers()
+			timer.Reset(d.resyncInterval)
+		}
+	}
+}
+
+func (d *demux) resyncLaggingWatchers() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	for w, nextRev := range d.laggingWatchers {
+		missedEvents := d.history.Filter(AfterRev(nextRev), w.predicate)
+		var latest int64
+		if latestEvent := d.history.PeekLatest(); latestEvent != nil {
+			latest = latestEvent.Kv.ModRevision
+		}
```

Comment:
You cannot access the history twice here. Each access locks history and provides atomic response, calling it twice means the data can be changed between. 

High level idea is correct, you want to return set nextRev to latest observed Rev even if it was filtered out by predicate. However you need to do atomically, so it should be returned as part of `Filter`.

Reply:
Short term I would recommend skip `PeekLatest`, it's just optimization. For now predicate should not filter anything so we can leave a TODO to optimize it. Having predicate here is premature optimization. I would recommend to just remove usage of predicate, to remove edge cases, simplify code and tests. It can always be added in the future.

Long term solution is pretty tricky, we can discuss it on the meeting when we merge this.